builder/chroots.go

```diff
@@ -0,0 +1,637 @@
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	w := bufio.NewWriter(f)
+	for _, e := range versions {
+		// TODO: change users of "versions" file to not rely on this exact formatting (version
+		// starting at column 51). E.g. this doesn't handle very well packages with large names.
+		fmt.Fprintf(w, "%-50s%s\n", e.name, e.version)
+	}
+	return w.Flush()
+}
+
+func fixOSRelease(filename, version string) error {
+	f, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	// TODO: If this is a mix, NAME and ID should probably change too. Create a section in
+	// configuration that will be used as reference to fill this.
+
+	// TODO: If this is a mix, add extra field for keeping track of the Clear Linux version
+	// used. Maybe also put the UPSTREAM URL, so we are ready to support mixes of mixes.
+
+	var newBuf bytes.Buffer
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		text := scanner.Text()
+		if strings.HasPrefix(text, "VERSION_ID=") {
+			text = "VERSION_ID=" + version
+		}
+		fmt.Fprintln(&newBuf, text)
+	}
+
+	err = scanner.Err()
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(filename, newBuf.Bytes(), 0644)
+}
+
+type osInfo struct {
+	ID        string
+	VersionID string
+}
+
+// readOSInfo reads the os-release(5) file to collect information about the system.
+func readOSInfo() (*osInfo, error) {
+	f, err := os.Open("/etc/os-release")
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		f, err = os.Open("/usr/lib/os-release")
+		if err != nil {
+			return nil, err
+		}
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	release, err := ini.Load(f)
+	if err != nil {
+		return nil, err
+	}
+	section := release.Section("")
+
+	info := &osInfo{
+		ID:        section.Key("ID").Value(),
+		VersionID: section.Key("VERSION_ID").Value(),
+	}
+	return info, nil
+}
+
+func installPackagesToBundleChroot(yumCmd []string, chrootVersionDir string, bundle *bundle) error {
+	baseDir := filepath.Join(chrootVersionDir, bundle.Name)
+	args := merge(yumCmd,
+		"--installroot="+baseDir,
+		"install",
+	)
+	args = append(args, bundle.AllPackages...)
+	_, err := run(args...)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filepath.Join(baseDir, "usr/share/clear/bundles", bundle.Name), nil, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Generate packages-{BUNDLE} file, that contains the list of packages and package versions
+	// present in each bundle.
+	packages, err := run(
+		"rpm",
+		"--root="+baseDir,
+		"-qa",
+		"--queryformat", "%{NAME}\t%{SOURCERPM}\n",
+	)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(filepath.Join(chrootVersionDir, "packages-"+bundle.Name), packages.Bytes(), 0644)
+}
+
+// run executes args storing its output in memory. If the command succeeds returns that output, if
+// it fails, return err that contains both the out and err streams from the execution.
+func run(args ...string) (*bytes.Buffer, error) {
```

Comment:
Put this in helpers.go where RunCommand and RunCommandSilent are? Or update one/both of them to do what you need here? Don't want 3 functions doing essentially the same thing in different files.

Reply:
This should definitely be de-duplicated.

Reply:
Done. Make total sense, they just weren't in the codebase when the PR was made. :)

Created RunCommandOutput, which does what the run here does, and implemented RunCommandSilent using that. Added two tests for RunCommandOutput checking the output and error message have what we expect.